dhis-2/dhis-support/dhis-support-hibernate/src/main/java/org/hisp/dhis/hibernate/HibernateGenericStore.java

```diff
@@ -575,6 +576,19 @@ public int getCount()
         return objects.isEmpty() || (object != null && objects.size() == 1 && object.equals( objects.get( 0 ) ) );
     }
 
+    @Override
+    public List<T> getAllByAttributeAndValues( Attribute attribute, List<String> values )
+    {
+        CriteriaBuilder builder = getCriteriaBuilder();
+
+        CriteriaQuery<T> query = builder.createQuery( getClazz() );
+        Root<T> root = query.from( getClazz() );
+        query.select( root );
+        query.where( builder.function( FUNCTION_JSONB_EXTRACT_PATH_TEXT, String.class, root.get( "attributeValues" ), builder.literal( attribute.getUid() ), builder.literal( "value" ) ).in( values ) );

```

Comment:
Have you done any performance testing here? should we also look into adding jsonb indexes?

Reply:
Last time I couldn't find away to add index, reasons is :

- GIN index doesn't support those functions `jsonb_extract_path_text()` , it only supports  jsonb operators  like `@>`
- Codes in `HibernateGenericStore`  need to be generic, so we can't use native sql with jsonb operators

- JPA  query API doesn't support jsonb operators, only postgresql built-in functions like `jsonb_extract_path_text()`, and it also need to be registered in `DhisPostgresDialect`

`registerFunction( "jsonb_extract_path_text", new StandardSQLFunction("jsonb_extract_path_text", StandardBasicTypes.STRING ));`

Reply:
This is based directly on tehe same query used for attributes in generl :) I think the main performance improvement here would be indexes, which we can look into in the future. It seems like @vietnguyen has already looked into indexes, so maybe we need to wait for better support, or rethink how we query the data.

Reply:
Ok @vietnguyen I see, we should look into it later also... PostgreSQL keeps getting updated and we might at some point be able to use it (`@>` is course also available as a function you can use, but i guess thats not really the problem)